lib/charms/opensearch/v0/opensearch_locking.py

```diff
@@ -343,7 +344,12 @@ def release(self):
         if host or alt_hosts:
             logger.debug("[Node lock] Checking which unit has opensearch lock")
             # Check if this unit currently has lock
-            if self._unit_with_lock(host) == self._charm.unit.name:
+            # or if there is a stale lock from a unit no longer existing
+            if (
+                self._unit_with_lock(host) == self._charm.unit.name
+                or self._unit_with_lock(host)
+                not in self._charm.model.get_relation(self._ENDPOINT_NAME).units
```

Comment:
hey @reneradoi, you do not want to do:
```
self._unit_with_lock(host) not in self._charm.model.get_relation(self._ENDPOINT_NAME).units
```

OpenSearch clusters may be composed of several `juju applications`. This is the *large deployments* scenario, where we have each `juju app` running different node roles.
So, that endpoint will not always have all nodes available.

I'd recommend instead:
```
                self._unit_with_lock(host) == self._charm.unit.name
                or (
                      self.opensearch_peer_cm.deployment_desc().typ == DeploymentType.MAIN_ORCHESTRATOR
                      and self._charm.unit.is_leader()
                      and self._unit_with_lock(host)
                      not in self._A_UNIT_LIST
                )
```

Now, `self._A_UNIT_LIST` should be something similar to:
```
[self._charm.unit]
 + [unit for unit in rel.units for rel in self._charm.model.get_relation("peer-cluster-orchestrator")]
 + [unit for unit in rel.units for rel in self._charm.model.get_relation("peer-cluster")]
 + self._charm.model.get_relation(self._ENDPOINT_NAME).units
```

@Mehdi-Bendriss likely has simpler way to build this list above.

Reply:
> OpenSearch clusters may be composed of several `juju applications`. This is the _large deployments_ scenario, where we have each `juju app` running different node roles.
> So, that endpoint will not always have all nodes available.

Good point, didn't think about that!

Reply:
Testing this manually (with the setup described in the issue) and with the integration tests in CI pipeline was successfull. Please let me know if this solution is appropriate.

Reply:
This is a bit tricky in the case of large deployments, as we don't have visibility over non-online units in other applications - which may be holding the lock.
it should be around:
```
unit_with_lock = self._unit_with_lock(host)

current_app_units = [u.name for u in charm.model.get_relation(PeerRelationName).units]
other_apps_units = ?? # for large deployments, we don't have this mechanism yet - for units that are not online

if (
    unit_with_lock
    and (
        unit_with_lock == self._charm.unit.name
        or unit_with_lock not in current_app_units
        or unit_with_lock not in other_apps_units
    )
)
```

I think that for large deployments we're missing the `other_apps_units` part

Reply:
Tracked here https://github.com/canonical/opensearch-operator/issues/314